### PR TITLE
T005 store shop items in db

### PR DESCRIPTION
**Pull request by vvs-code**
Store items in db and add KnpLabs/KnpPaginatorBundle

### CODE REVIEW

File: src/Shop/WebSiteBundle/Entity/Product.php [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+    /**
+     * @inheritDoc
+     */
+    public function unserialize($serialized)
+    {
+        $arr = \json_decode($serialized);
+        foreach ($arr as $key => $val) {
+            $this->$key = $val;
+        }
+        $this->categories = \json_decode($this->categories);
```
Comment (vvs-code) [inline src/Shop/WebSiteBundle/Entity/Product.php]: Fixed

File: src/Shop/WebSiteBundle/Entity/Product.php [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+     */
+    public function serialize()
+    {
+        return \json_encode([
+            $this->id,
+            $this->image,
+            $this->title,
+            $this->price,
+            $this->description,
+            json_encode($this->categories)
```
Comment (vvs-code) [inline src/Shop/WebSiteBundle/Entity/Product.php]: Done

File: app/DoctrineMigrations/Version20141221083545.php [hunk 3]
```diff
@@ -0,0 +1,28 @@
+<?php
```
Comment (vvs-code) [inline app/DoctrineMigrations/Version20141221083545.php]: Ok

File: src/Shop/WebSiteBundle/Entity/ProductRepository.php [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+use Shop\CommonBundle\Entity\CommonRepository;
+use Knp\Component\Pager\Pagination\PaginationInterface;
+use Knp\Component\Pager\Paginator;
+
+class ProductRepository extends CommonRepository
+{
+    /**
+     * @var Paginator KnpPaginator
+     */
+    protected $paginator;
```
Comment (vvs-code) [inline src/Shop/WebSiteBundle/Entity/ProductRepository.php]: Done

File: src/Shop/WebSiteBundle/Entity/ProductRepository.php [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    public function setPaginator(Paginator $paginator)
+    {
+        $this->paginator = $paginator;
+    }
+
+    /**
+     * Return paganation for passed params
+     * @return PaginationInterface
+     */
+    public function getPagination($page = 1, $limit = 10, array $options = [])
```
Comment (vvs-code) [inline src/Shop/WebSiteBundle/Entity/ProductRepository.php]: OK

File: src/Shop/WebSiteBundle/Resources/config/config.yml [hunk 6]
```diff
@@ -1,7 +1,22 @@
...
 
 services:
+  shop.website.product_repository:
+      class: %shop.website.product_repository%
+      factory_service: doctrine.orm.entity_manager
+      factory_method:  getRepository
+      arguments:
+          - 'Shop\WebSiteBundle\Entity\Product'
+      calls:
+        - [setPaginator, ["@knp_paginator"]]
```
Comment (vvs-code) [inline src/Shop/WebSiteBundle/Resources/config/config.yml]: Done

File: src/Shop/WebSiteBundle/Service/ProductService.php [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    public function createProduct(array $properties = [])
+    {
+        /**
+         * @var Product
+         */
+        $product = new Product();
+        /**
+         * @var \ReflectionClass
+         */
+        $reflection = new \ReflectionClass($product);
```
Comment (vvs-code) [inline src/Shop/WebSiteBundle/Service/ProductService.php]: Fixed

File: src/Shop/WebSiteBundle/Service/ProductServiceInterface.php [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+use Knp\Component\Pager\Pagination\PaginationInterface;
+
+interface ProductServiceInterface
+{
+    /**
+     * Create product
+     * @param array $properties
+     * @return Product
+     */
+    public function createProduct(array $properties);
```
Comment (vvs-code) [inline src/Shop/WebSiteBundle/Service/ProductServiceInterface.php]: added

File: src/Shop/WebSiteBundle/Service/ProductService.php [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+     */
+    public function findBy(array $criteria = [], array $orderBy = null, $limit = null, $offset = null)
+    {
+        return $this->productRepository->findBy($criteria, $orderBy, $limit, $offset);
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function getPagination($page = 1, $limit = 10, array $options =[])
```
Comment (vvs-code) [inline src/Shop/WebSiteBundle/Service/ProductService.php]: add

File: src/Shop/WebSiteBundle/Controller/IndexController.php [hunk 10]
```diff
@@ -3,22 +3,31 @@
...
+use Shop\WebSiteBundle\Service\ProductService;
+use Symfony\Component\HttpFoundation\Request;
+
 
 /**
  * @Route(service="shop.website.index_controller")
  */
 class IndexController extends Controller\CommonController
 {
+    const QUERY_PAGE_KEY = 'page';
```
Comment (vvs-code) [inline src/Shop/WebSiteBundle/Controller/IndexController.php]: i think setter a more flexible - no need to change constructor signature

File: src/Shop/WebSiteBundle/Controller/IndexController.php [hunk 11]
```diff
@@ -29,4 +43,17 @@ public function cartAction()
         return $this->render('WebSiteBundle:Index:cart.html.twig');
     }
 
+    public function setProductService(ProductService $productService)
```
Comment (EduardTrayan) [inline src/Shop/WebSiteBundle/Controller/IndexController.php]: Please add docblock

File: src/Shop/WebSiteBundle/Entity/Product.php [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+    /**
+     * @ORM\Column(name="description", type="text")
+     *
+     * @var string $title
+     */
+    private $description;
+    /**
+     * @ORM\Column(name="price", type="integer")
+     *
+     * @var string $price
```
Comment (EduardTrayan) [inline src/Shop/WebSiteBundle/Entity/Product.php]: I suppose type of this variable is `integer`

File: src/Shop/WebSiteBundle/Entity/Product.php [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+    /**
+     * @param string $image
+     */
+    public function setImage($image)
+    {
+        $this->image = $image;
+    }
+
+    /**
+     * @return string
```
Comment (EduardTrayan) [inline src/Shop/WebSiteBundle/Entity/Product.php]: I suppose type of return variable is `integer`

File: src/Shop/WebSiteBundle/Entity/Product.php [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+     */
+    public function getPrice()
+    {
+        return $this->price;
+    }
+
+    /**
+     * @param string $price
```
Comment (EduardTrayan) [inline src/Shop/WebSiteBundle/Entity/Product.php]: I suppose type of return variable is `integer`

File: src/Shop/WebSiteBundle/Controller/IndexController.php (hunk 10 above)
Comment (EduardTrayan) [inline src/Shop/WebSiteBundle/Controller/IndexController.php]: Ok, then inject parameter via setter

Comment (vvs-code) [inline src/Shop/WebSiteBundle/Controller/IndexController.php]: Done

File: src/Shop/WebSiteBundle/Entity/Product.php (hunk 12 above)
Comment (vvs-code) [inline src/Shop/WebSiteBundle/Entity/Product.php]: Fixed

File: src/Shop/WebSiteBundle/Entity/Product.php (hunk 13 above)
Comment (vvs-code) [inline src/Shop/WebSiteBundle/Entity/Product.php]: Fixed

File: src/Shop/WebSiteBundle/Entity/Product.php (hunk 14 above)
Comment (vvs-code) [inline src/Shop/WebSiteBundle/Entity/Product.php]: Fixed

File: src/Shop/WebSiteBundle/Controller/IndexController.php [hunk 15]
```diff
@@ -3,22 +3,36 @@
 
 use Shop\CommonBundle\Controller;
 use Sensio\Bundle\FrameworkExtraBundle\Configuration\Route;
+use Shop\WebSiteBundle\Entity\Product;
```
Comment (EduardTrayan) [inline src/Shop/WebSiteBundle/Controller/IndexController.php]: Please check order of dependencies (should be in alphabetical order)

File: src/Shop/WebSiteBundle/Controller/IndexController.php (hunk 10 above)
Comment (EduardTrayan) [inline src/Shop/WebSiteBundle/Controller/IndexController.php]: I don't see any changes

Comment (vvs-code) [inline src/Shop/WebSiteBundle/Controller/IndexController.php]: Sorry i miss first comment sense. Fixed ( moved to main config )

File: app/Resources/views/macros.html.twig [hunk 16]
```diff
@@ -41,6 +41,6 @@
 {% endmacro %}
 
 {% macro printItem(item) %}
-    <img src="{{ item.img }}" align="left"/><strong>{{ item.title }}</strong><button>Buy</button>
-    <p>{{ item.decriptions|raw }}</p>
+    <img src="{{ item.image }}" align="left"/><strong>{{ item.title }}</strong><button>Buy</button>
```
Comment (EduardTrayan) [inline app/Resources/views/macros.html.twig]: If your entity calls `product` please change name of variables which contain his entity

File: app/Resources/views/macros.html.twig [hunk 17]
```diff
@@ -41,6 +41,6 @@
 {% endmacro %}
 
 {% macro printItem(item) %}
```
Comment (EduardTrayan) [inline app/Resources/views/macros.html.twig]: If your entity calls `product` please change name of variables which contain his entity

File: app/Resources/views/macros.html.twig [hunk 18]
```diff
@@ -41,6 +41,6 @@
 {% endmacro %}
 
 {% macro printItem(item) %}
-    <img src="{{ item.img }}" align="left"/><strong>{{ item.title }}</strong><button>Buy</button>
-    <p>{{ item.decriptions|raw }}</p>
+    <img src="{{ item.image }}" align="left"/><strong>{{ item.title }}</strong><button>Buy</button>
+    <p>{{ item.description|raw }}</p>
```
Comment (EduardTrayan) [inline app/Resources/views/macros.html.twig]: If your entity calls `product` please change name of variables which contain his entity

File: app/config/config.yml [hunk 19]
```diff
@@ -2,6 +2,9 @@ imports:
     - { resource: parameters.yml }
     - { resource: security.yml }
 
+parameters:
+    page_query_key: page
```
Comment (EduardTrayan) [inline app/config/config.yml]: :+1:

File: src/Shop/WebSiteBundle/Resources/views/Index/index.html.twig [hunk 20]
```diff
@@ -2,12 +2,14 @@
 
 {% block title %}Main page{% endblock %}
 {% block body %}
-    <ul class="items">
-        {% for item in items %}
+    <ul class="pagination">
+        {% for item in pagination %}
```
Comment (EduardTrayan) [inline src/Shop/WebSiteBundle/Resources/views/Index/index.html.twig]: If your entity calls `product` please change name of variables which contain his entity

Comment (vvs-code) [inline src/Shop/WebSiteBundle/Resources/views/Index/index.html.twig]: Done

File: app/Resources/views/macros.html.twig (hunk 18 above)
Comment (vvs-code) [inline app/Resources/views/macros.html.twig]: Done

File: app/Resources/views/macros.html.twig (hunk 17 above)
Comment (vvs-code) [inline app/Resources/views/macros.html.twig]: Done

File: app/Resources/views/macros.html.twig (hunk 16 above)
Comment (vvs-code) [inline app/Resources/views/macros.html.twig]: Done

**Comment by EduardTrayan** [comment]
:+1:

**Pull request merged**